arlas-core/src/main/java/io/arlas/server/services/ExploreServices.java

```diff
@@ -350,25 +350,38 @@ protected void applyProjection(Projection projection, FluidSearch fluidSearch, O
         }
     }
 
-    public AggregationResponse formatAggregationResult(MultiBucketsAggregation aggregation, AggregationResponse aggregationResponse, String collection) {
+    public AggregationResponse formatAggregationResult(MultiBucketsAggregation aggregation, AggregationResponse aggregationResponse, CollectionReference collection, List<Aggregation> aggregationsRequest, int aggTreeDepth) {
         aggregationResponse.name = aggregation.getName();
         if (aggregationResponse.name.equals(FluidSearch.TERM_AGG)) {
             aggregationResponse.sumotherdoccounts = ((Terms) aggregation).getSumOfOtherDocCounts();
         }
+        RawGeometries rawGeometries = aggregationsRequest.size() > aggTreeDepth ? aggregationsRequest.get(aggTreeDepth).rawGeometries : null;
+        List<AggregatedGeometryEnum> aggregatedGeometries = aggregationsRequest.size() > aggTreeDepth ? aggregationsRequest.get(aggTreeDepth).aggregatedGeometries : null;
         aggregationResponse.elements = new ArrayList<AggregationResponse>();
         List<MultiBucketsAggregation.Bucket> buckets = (List<MultiBucketsAggregation.Bucket>) aggregation.getBuckets();
         buckets.forEach(bucket -> {
             AggregationResponse element = new AggregationResponse();
             element.keyAsString = bucket.getKeyAsString();
-            if (aggregationResponse.name.startsWith(FluidSearch.GEOHASH_AGG)) {
+            /** if it is a `geohash` aggregation type, we set the GEOHASHCENTER and GEOHASH aggregated geometries if they're requested **/
+            if (aggregationResponse.name.equals(FluidSearch.GEOHASH_AGG)) {
                 GeoPoint geoPoint = getGeohashCentre(element.keyAsString.toString());
                 element.key = geoPoint;
-                if (aggregationResponse.name.equals(FluidSearch.GEOHASH_AGG)) {
-                    // return the centroid of the geohash
-                    element.geometry = new Point(geoPoint.getLon(), geoPoint.getLat());
-                } else {
-                    // return the Extent of the geohash
-                    element.geometry = createPolygonFromRectangle(GeohashUtils.decodeBoundary(element.keyAsString.toString(), SpatialContext.GEO));
+                if (!CollectionUtils.isEmpty(aggregatedGeometries)) {
+                    aggregatedGeometries.stream().filter(g -> g == AggregatedGeometryEnum.GEOHASHCENTER || g == AggregatedGeometryEnum.GEOHASH).forEach(g -> {
+                        ReturnedGeometry returnedGeometry = new ReturnedGeometry();
+                        returnedGeometry.reference = g.value();
+                        returnedGeometry.isRaw = false;
+                        if (g == AggregatedGeometryEnum.GEOHASH) {
```

Comment:
I feel like there are multiple `if` with a lot of code everytime, could each case be put in a dedicated method to male this method easier to read, and with less cyclomatic complexity?

Reply:
I agree, but I think I should treat it another PR